Makefile

```diff
@@ -2,37 +2,148 @@
+
+##@ Environment
+# We use a volume here as non-Linux hosts are extremely slow to share disks, and Linux hosts tend to get permissions clobbered.
+define ENVIRONMENT_EXEC
+	@echo "Entering environment..."
+	@mkdir -p target
+	$(CONTAINER_TOOL) run \
+			--name vector-environment \
+			--rm \
+			$(if $(findstring true,$(ENVIRONMENT_TTY)),--tty,) \
+			--init \
+			--interactive \
+			--env INSIDE_ENVIRONMENT=true \
+			--network host \
+			--mount type=bind,source=${PWD},target=/vector \
+			--mount type=bind,source=/var/run/docker.sock,target=/var/run/docker.sock \
+			--mount type=volume,source=vector-target,target=/vector/target \
+			--mount type=volume,source=vector-cargo-cache,target=/root/.cargo \
+			$(ENVIRONMENT_UPSTREAM)
+endef
+
+define ENVIRONMENT_COPY_ARTIFACTS
+	@echo "Copying artifacts off volumes... (Docker errors below are totally okay)"
+	@mkdir -p ./target/release
+	@mkdir -p ./target/debug
+	@mkdir -p ./target/criterion
+	@$(CONTAINER_TOOL) rm -f vector-build-outputs || true
+	@$(CONTAINER_TOOL) run \
+		-d \
+		-v vector-target:/target \
+		--name vector-build-outputs \
+		busybox true
+	@$(CONTAINER_TOOL) cp vector-build-outputs:/target/release/vector ./target/release/ || true
+	@$(CONTAINER_TOOL) cp vector-build-outputs:/target/debug/vector ./target/debug/ || true
+	@$(CONTAINER_TOOL) cp vector-build-outputs:/target/criterion ./target/criterion || true
+	@$(CONTAINER_TOOL) rm -f vector-build-outputs
+endef
+
+
+ifeq ($(ENVIRONMENT_AUTOBUILD), true)
+define ENVIRONMENT_PREPARE
+	@echo "Building the environment. (ENVIRONMENT_AUTOBUILD=true) This may take a few minutes..."
+	$(CONTAINER_TOOL) build \
+		$(if $(findstring true,$(VERBOSE)),,--quiet) \
+		--tag $(ENVIRONMENT_UPSTREAM) \
+		--file scripts/environment/Dockerfile \
+		.
+endef
+else
+define ENVIRONMENT_PREPARE
+	$(CONTAINER_TOOL) pull $(ENVIRONMENT_UPSTREAM)
+endef
+endif
+
 
-##@ Default
+environment: ## Enter a full Vector dev shell in $CONTAINER_TOOL, binding this folder to the container.
+	${ENVIRONMENT_PREPARE}
+	@export ENVIRONMENT_TTY=true
+	${ENVIRONMENT_EXEC}
 
-all: check build-all package-all test-docker test-behavior verify ## Run all tests, checks, and verifications
+environment-prepare: ## Prepare the Vector dev shell using $CONTAINER_TOOL.
+	${ENVIRONMENT_PREPARE}
+
+environment-clean: ## Clean the Vector dev shell using $CONTAINER_TOOL.
+	@$(CONTAINER_TOOL) volume rm -f vector-target vector-cargo-cache
+	@$(CONTAINER_TOOL) rmi $(ENVIRONMENT_UPSTREAM) || true
+
+environment-push: environment-prepare ## Publish a new version of the container image.
+	$(CONTAINER_TOOL) push $(ENVIRONMENT_UPSTREAM)
 
 ##@ Building
+build: ## Build the project in release mode (Supports `ENVIRONMENT=true`)
+ifeq ($(ENVIRONMENT), true)
+	${ENVIRONMENT_PREPARE}
+	${ENVIRONMENT_EXEC} make build
+	${ENVIRONMENT_COPY_ARTIFACTS}
+else
+	cargo build --release --no-default-features --features ${DEFAULT_FEATURES}
+endif
```

Comment:
You can probably use a canned recipe to DRY these:
```make
ifeq ($(ENVIRONMENT), true)
define magic =
${ENVIRONMENT_PREPARE}
${ENVIRONMENT_EXEC} make $@
endef
else
define magic =
$1
endef
endif

build:
        $(call magic,"cargo build --release --no-default-features --features ${DEFAULT_FEATURES}")
```

Of course, you would need separate macros where one did the `${ENVIRONMENT_COPY_ARTIFACTS}`, and it would probably get ugly for multi-line recipes. Another possibility would be defining a sub-rule, which would allow for multi-line recipes:
```make
ifeq ($(ENVIRONMENT), true)
define magic = 
${ENVIRONMENT_PREPARE}
${ENVIRONMENT_EXEC} make REAL_$@
endef
else
define magic =
@$(MAKE) REAL_$@
endef
endif

build:
        $(magic)
REAL_build:
        cargo build --release --no-default-features --features ${DEFAULT_FEATURES}
```

Reply:
You're totally right, and I considered making it work like our `RUN` val right now... However I also noted as you did that some jobs would require special concerns like moving files etc, so I opted not to. I tried to balance reusability and understanding, maybe I lost. :)

Let me try to incorporate this feedback. :) Will report back.

Reply:
![image](https://user-images.githubusercontent.com/130903/84417467-85843900-abca-11ea-9d95-8d61ff76e13c.png)
![image](https://user-images.githubusercontent.com/130903/84417470-887f2980-abca-11ea-9dda-abf55858669e.png)


Much better